cmd/explaintest/r/topn_push_down.result

```diff
@@ -218,3 +218,56 @@ Apply_17	9990.00	root	semi join, inner:Selection_21, equal:[eq(test.t1.a, x.a)]
         └─Limit_28	1.00	cop	offset:0, count:1
           └─Selection_27	1.00	cop	gt(test.t2.b, test.t1.b)
             └─TableScan_26	1.25	cop	table:t2, range:[-inf,+inf], keep order:false, stats:pseudo
+drop table if exists t;
+create table t(a int not null, index idx(a));
+explain select /*+ TIDB_INLJ(t2) */ * from t t1 join t t2 on t1.a = t2.a limit 5;
+id	count	task	operator info
+Limit_11	5.00	root	offset:0, count:5
+└─IndexJoin_15	5.00	root	inner join, inner:IndexReader_14, outer key:t1.a, inner key:t2.a
+  ├─TableReader_17	4.00	root	data:TableScan_16
+  │ └─TableScan_16	4.00	cop	table:t1, range:[-inf,+inf], keep order:false, stats:pseudo
+  └─IndexReader_14	10.00	root	index:IndexScan_13
+    └─IndexScan_13	10.00	cop	table:t2, index:a, range: decided by [t1.a], keep order:false, stats:pseudo
+explain select /*+ TIDB_INLJ(t2) */ * from t t1 left join t t2 on t1.a = t2.a where t2.a is null limit 5;
+id	count	task	operator info
+Limit_12	5.00	root	offset:0, count:5
+└─Selection_13	5.00	root	isnull(t2.a)
+  └─IndexJoin_17	5.00	root	left outer join, inner:IndexReader_16, outer key:t1.a, inner key:t2.a
+    ├─TableReader_19	5.00	root	data:TableScan_18
+    │ └─TableScan_18	5.00	cop	table:t1, range:[-inf,+inf], keep order:false, stats:pseudo
+    └─IndexReader_16	10.00	root	index:IndexScan_15
+      └─IndexScan_15	10.00	cop	table:t2, index:a, range: decided by [t1.a], keep order:false, stats:pseudo
+explain select /*+ TIDB_SMJ(t1, t2) */ * from t t1 join t t2 on t1.a = t2.a limit 5;
+id	count	task	operator info
+Limit_11	5.00	root	offset:0, count:5
+└─MergeJoin_12	5.00	root	inner join, left key:t1.a, right key:t2.a
+  ├─IndexReader_14	4.00	root	index:IndexScan_13
+  │ └─IndexScan_13	4.00	cop	table:t1, index:a, range:[NULL,+inf], keep order:true, stats:pseudo
+  └─IndexReader_16	4.00	root	index:IndexScan_15
+    └─IndexScan_15	4.00	cop	table:t2, index:a, range:[NULL,+inf], keep order:true, stats:pseudo
+explain select /*+ TIDB_SMJ(t1, t2) */ * from t t1 left join t t2 on t1.a = t2.a where t2.a is null limit 5;
+id	count	task	operator info
+Limit_12	5.00	root	offset:0, count:5
+└─Selection_13	5.00	root	isnull(t2.a)
+  └─MergeJoin_14	5.00	root	left outer join, left key:t1.a, right key:t2.a
+    ├─IndexReader_16	5.00	root	index:IndexScan_15
+    │ └─IndexScan_15	5.00	cop	table:t1, index:a, range:[NULL,+inf], keep order:true, stats:pseudo
+    └─IndexReader_18	10000.00	root	index:IndexScan_17
+      └─IndexScan_17	10000.00	cop	table:t2, index:a, range:[NULL,+inf], keep order:true, stats:pseudo
```

Comment:
@zz-jason @lamxTyler Here is another problem I am not sure: how should we set expected count for inner child of InnerMergeJoin?

Reply:
What about set it according to the formula in `DeriveStats` of join?

Reply:
The formulas in `LogicalJoin::DeriveStats` mainly use row count of outer child for computing row count of join, except for InnerJoin / LeftOuterJoin / RightOuterJoin, then we would end up with 4.00 as expected count of both children plans for inner join or left outer join? how about inner child of semi joins?

Reply:
How about this simple strategy?
Since we only need `ExpectedCnt` rows from `outer`, the estimated cardinality of outer's result is `ExpectedCnt * outer.Cardinality / outer.rowCount`.
And the cardinality scale of inner's result is `inner.rowCount / inner.Cardinality`.
So in order to get all keys which appear in outer's result from inner, we have to fetch `(ExpectedCnt * outer.Cardinality / outer.rowCount) * (inner.rowCount / inner.Cardinality)` rows from inner, which should be inner's `ExpectedCnt`?